src/client/auth/x509.rs

```diff
@@ -58,7 +58,7 @@ pub(super) async fn authenticate_stream(
             .auth_response_body("MONGODB-X509")?,
     };
 
-    if server_response.get_str("dbname") != Ok("$external") {
+    if server_response.get_i32("ok") != Ok(1) {
```

Comment:
sometimes the server can return different types for this integer (and sometimes even a double), so we use `bson_util::get_int` for these kinds of checks to ensure all types are covered.

Reply:
It turns out that `ok: 1` is itself not always present, and apparently the thing to do is just not inspect the server response at all (i.e. https://github.com/mongodb/mongo-python-driver/blob/89f41cfbd2560b41592dadf6e1b59f80c33ec207/pymongo/auth.py#L424).

Reply:
Every command should always return an `ok` value, so that sounds like a bug in the server / proxy. What circumstances could trigger that? (I think in the Python example, the `command` method probably will throw an exception if `ok: 0` is encountered). I think we need to check _some_ field here to properly determine if authentication worked or not.

Reply:
This is observed behavior in the Evergreen x509 tests (i.e. [here](https://evergreen.mongodb.com/lobster/evergreen/task/mongo_rust_driver_x509_auth__async_runtime~tokio_auth_and_tls~auth_and_tls_os~macos_10.14_test_x509_auth_patch_f93fb4057b30490d9fb407b23451daa37646c6c4_61aa31932a60ed5e66cfdcbb_21_12_03_15_02_47/0/task#bookmarks=0%2C1413&shareLine=1105)); it looks like those are running against mongodb 4.4 in a standalone configuration.

Updated the check to look for either `ok: 1` or `dbname: $external`.

Reply:
Ah I think I understand now. The `ok` is omitted when the authentication is performed speculatively (i.e. we include the first step of authentication in our original handshake command, and the server includes their response to it in their initial handshake response). This happens because the authentication response is returned as a subdocument of a larger command response, and we extract it and pass it down through `authenticate_stream`. I see now that in the Python example, it returns early in that case. I think we can take a similar approach here to Python: if speculative auth happened (i.e. we have a server first value), just return `Ok(())`. Otherwise, do the next round and ensure `ok: 1`.

Reply:
Ah, I see, thanks.  Updated.